### PR TITLE
MINOR: Cleanup redundancies in BaseRequestTest

### PR DESCRIPTION
**Pull request by hachikuji**
This patch eliminates a lot of redundancy and general messiness around the usage of `BaseRequestTest` and specifically response deserialization.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/test/scala/unit/kafka/server/BaseRequestTest.scala [hunk 1]
```diff
@@ -84,104 +83,46 @@ abstract class BaseRequestTest extends IntegrationTestHarness {
...
+    ResponseHeader.parse(responseBuffer, apiKey.responseHeaderVersion(version))
+
+    val responseStruct = apiKey.parseResponse(version, responseBuffer)
+    AbstractResponse.parseResponse(apiKey, responseStruct, version)
   }
 
-  def requestAndReceive(socket: Socket, request: Array[Byte]): Array[Byte] = {
-    sendRequest(socket, request)
-    receiveResponse(socket)
+  def sendAndReceive(request: AbstractRequest, socket: Socket): AbstractResponse = {
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/server/BaseRequestTest.scala]: You can add a type parameter for the desired response type along with a ClassTag so that you can pattern match safely and either return the desired type or throw an exception. Then the callers no longer need to cast.

Comment (hachikuji) [reply core/src/test/scala/unit/kafka/server/BaseRequestTest.scala]: Thanks for the suggestion. Updated.

File: core/src/test/scala/integration/kafka/network/DynamicConnectionQuotaTest.scala [hunk 2]
```diff
@@ -100,8 +100,7 @@ class DynamicConnectionQuotaTest extends BaseRequestTest {
       val socket = connect("PLAINTEXT")
       socket.setSoTimeout(1000)
       try {
-        val response = sendAndReceive(produceRequest, ApiKeys.PRODUCE, socket)
-        assertEquals(0, response.remaining)
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/network/DynamicConnectionQuotaTest.scala:104]: Why don't we need this assertion anymore?

File: core/src/test/scala/unit/kafka/server/AbstractApiVersionsRequestTest.scala [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package kafka.server
+
+import org.apache.kafka.common.protocol.ApiKeys
+import org.apache.kafka.common.requests.{ApiVersionsRequest, ApiVersionsResponse}
+
+abstract class AbstractApiVersionsRequestTest extends BaseRequestTest {
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/server/AbstractApiVersionsRequestTest.scala:26]: Is this abstract base class worth it? We could either add this single method to `BaseRequestTest` or `TestUtils`.

File: core/src/test/scala/unit/kafka/server/AbstractCreateTopicsRequestTest.scala [hunk 4]
```diff
@@ -152,15 +151,6 @@ class AbstractCreateTopicsRequestTest extends BaseRequestTest {
   protected def error(error: Errors, errorMessage: Option[String] = None): ApiError =
     new ApiError(error, errorMessage.orNull)
 
-  protected def toStructWithDuplicateFirstTopic(request: CreateTopicsRequest): Struct = {
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/server/AbstractCreateTopicsRequestTest.scala:155]: Interesting that this is no longer used. Do you know if we are still testing the case where we have a duplicate topic?

File: core/src/test/scala/unit/kafka/server/ApiVersionsRequestTest.scala [hunk 5]
```diff
@@ -65,21 +65,20 @@ class ApiVersionsRequestTest extends BaseRequestTest {
   @Test
   def testApiVersionsRequestValidationV0(): Unit = {
     val apiVersionsRequest = new ApiVersionsRequest.Builder().build( 0.asInstanceOf[Short])
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/server/ApiVersionsRequestTest.scala]: Existing issue, but empty space before `0` should be removed.

File: core/src/test/scala/unit/kafka/server/DeleteTopicsRequestTest.scala [hunk 6]
```diff
@@ -131,12 +131,11 @@ class DeleteTopicsRequestTest extends BaseRequestTest {
   }
 
   private def sendDeleteTopicsRequest(request: DeleteTopicsRequest, socketServer: SocketServer = controllerSocketServer): DeleteTopicsResponse = {
-    val response = connectAndSend(request, ApiKeys.DELETE_TOPICS, socketServer)
-    DeleteTopicsResponse.parse(response, request.version)
+    connectAndReceive[DeleteTopicsResponse](request, destination = socketServer)
   }
 
   private def sendMetadataRequest(request: MetadataRequest): MetadataResponse = {
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/server/DeleteTopicsRequestTest.scala]: Is it even worth having this method? There are a few other examples like that in other classes.

File: core/src/test/scala/unit/kafka/server/RequestQuotaTest.scala [hunk 7]
```diff
@@ -561,80 +558,23 @@ class RequestQuotaTest extends BaseRequestTest {
     }
   }
 
-  private def responseThrottleTime(apiKey: ApiKeys, response: Struct): Int = {
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/server/RequestQuotaTest.scala:564]: Glad to see this go. I had done something similar in https://github.com/apache/kafka/pull/7409/files#diff-8259941614685619733d9f458762a0b6L564, but hadn't gotten around to pushing that PR over the line.

File: core/src/test/scala/unit/kafka/server/BaseRequestTest.scala [hunk 8]
```diff
@@ -84,107 +85,69 @@ abstract class BaseRequestTest extends IntegrationTestHarness {
...
   }
 
-  def nextRequestHeader(apiKey: ApiKeys, apiVersion: Short): RequestHeader = {
-    correlationId += 1
-    new RequestHeader(apiKey, apiVersion, "client-id", correlationId)
+  def nextRequestHeader[T <: AbstractResponse](apiKey: ApiKeys,
+                                               apiVersion: Short,
+                                               clientId: String = "client-id",
+                                               correlationIdOpt: Option[Int] = None
+                                              ): RequestHeader = {
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/server/BaseRequestTest.scala]: Nit: this is a bit unusual, any reason for it not to be on the previous line?

File: core/src/test/scala/unit/kafka/server/SaslApiVersionsRequestTest.scala [hunk 9]
```diff
@@ -52,7 +50,7 @@ class SaslApiVersionsRequestTest extends BaseRequestTest with SaslSetup {
 
   @Test
   def testApiVersionsRequestBeforeSaslHandshakeRequest(): Unit = {
-    val plaintextSocket = connect(protocol = securityProtocol)
+    val plaintextSocket = connect()
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/server/SaslApiVersionsRequestTest.scala]: Maybe we should rename this to `socket` as the current name is a bit misleading. Same for the other methods.

**Review by ijuma — COMMENTED** [review]
Thanks for the PR. Nice clean-up. A few minor comments.

File: core/src/test/scala/unit/kafka/server/AbstractApiVersionsRequestTest.scala (hunk 3 above)
Comment (hachikuji) [reply core/src/test/scala/unit/kafka/server/AbstractApiVersionsRequestTest.scala:26]: It's a bit more justifiable if we take the `validateApiVersionsResponse` method out of the `ApiVersionsRequestTest` companion.

Comment (ijuma) [reply core/src/test/scala/unit/kafka/server/AbstractApiVersionsRequestTest.scala:26]: Maybe we should move this method to the companion? Nice to avoid long inheritance hierarchies.

File: core/src/test/scala/integration/kafka/network/DynamicConnectionQuotaTest.scala (hunk 2 above)
Comment (hachikuji) [reply core/src/test/scala/integration/kafka/network/DynamicConnectionQuotaTest.scala:104]: The assertion was actually never getting executed. The call to `sendAndReceive` raises a timeout exception, which is what `verifyMaxConnections` is looking for.

File: core/src/test/scala/unit/kafka/server/AbstractCreateTopicsRequestTest.scala (hunk 4 above)
Comment (hachikuji) [reply core/src/test/scala/unit/kafka/server/AbstractCreateTopicsRequestTest.scala:155]: I tracked the removal of the uses of this method back to this: https://github.com/apache/kafka/commit/e2e8bdbd8cb6ca2ac962c72147d21a9e8b9ba2c0#diff-217dea05d3ff55c8bf10503c329f7c2b. I don't see a clear reason for it.

File: core/src/test/scala/unit/kafka/server/AbstractApiVersionsRequestTest.scala (hunk 3 above)
Comment (hachikuji) [reply core/src/test/scala/unit/kafka/server/AbstractApiVersionsRequestTest.scala:26]: We could but it's a little more awkward since it requires an instance of `BaseRequestTest`. We could also add it to `BaseRequestTest`, but I was hoping to keep that class generic and to avoid leaking a bunch of api-specific noise into it.

**Review by zhengb0 — APPROVED** [review]

**Comment by zhengb0** [comment]
I'm sorry, I'm not careful

File: core/src/test/scala/unit/kafka/server/AbstractCreateTopicsRequestTest.scala (hunk 4 above)
Comment (ijuma) [reply core/src/test/scala/unit/kafka/server/AbstractCreateTopicsRequestTest.scala:155]: @cmccabe do you remember why the test was removed?

**Review by ijuma — APPROVED** [review]
LGTM

**Comment by hachikuji** [comment]
retest this please

**Comment by hachikuji** [comment]
retest this please

File: core/src/test/scala/unit/kafka/server/AbstractCreateTopicsRequestTest.scala (hunk 4 above)
Comment (cmccabe) [reply core/src/test/scala/unit/kafka/server/AbstractCreateTopicsRequestTest.scala:155]: CreateTopicsRequest was the first RPC converted over to automatic generation, if I remember correctly.  I converted it while writing the generator.

I think when the automatic code generation was first developed, it didn't have support for multiple identical elements in the same map.  So it wasn't possible to serialize the invalid request at that point.  So I had to remove this test.  Later on I added support for duplicate entries in the same map.

We should definitely re-add a test for duplicate createTopics elements.  There should also be a test like this in KafkaAdminClientTest to test the admin client's handling of it (AC should not allow this to even get to the server)

**Comment by hachikuji** [comment]
retest this please

File: core/src/test/scala/unit/kafka/server/AbstractCreateTopicsRequestTest.scala (hunk 4 above)
Comment (ijuma) [reply core/src/test/scala/unit/kafka/server/AbstractCreateTopicsRequestTest.scala:155]: Can we file a JIRA please?
